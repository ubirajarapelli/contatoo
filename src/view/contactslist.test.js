/**
 * @jest-environment jsdom
 */

import fs from 'fs'
import path from 'path'
import fetch from 'jest-fetch-mock'
import '@testing-library/jest-dom/extend-expect'
import { fireEvent, getByText, getByTestId, getByRole, waitForElementToBeRemoved } from '@testing-library/dom'
import { JSDOM } from 'jsdom'
import ContactsListView from './contactsList.js'

const html = fs.readFileSync(path.resolve(__dirname, '../index.html'), 'utf8');

const mock = [
  {
    "name": "My name 1",
    "cpf": "04080757247",
    "phone": "11987654321",
    "email": "myemail1@test.com.br"
  },
  {
    "name": "My name 2",
    "cpf": "77797584192",
    "phone": "11987654321",
    "email": "myemail2@test.com.br"
  },
  {
    "name": "My name 3",
    "cpf": "45486737688",
    "phone": "11987654321",
    "email": "myemail3@test.com.br"
  },
  {
    "name": "My name 4",
    "cpf": "74668869066",
    "phone": "11987654321",
    "email": "myemail4@test.com.br"
  }
]

let options = {
  runScripts: 'dangerously',
}

let dom
let container
dom = new JSDOM(html, options)

const contactsListView = new ContactsListView
container = dom.window.document.body.innerHTML = contactsListView.$mainTemplate

describe('Unit tests from a Contact List',  () => {

  describe('When a load the contact`s list page', () => {
    test('Should render a empty list', ()  => {
      contactsListView.render()
      expect(container).toMatchSnapshot()
    })

    test('Should render with a inital data', ()  => {
      contactsListView.render(mock)
      expect(container).toMatchSnapshot()
    })
  })

  describe('When a contact`s list have contacts', () => {
    beforeEach(() => {
      contactsListView.render(mock)
    })

    test('Should click a exclude button and remove a contact', async () => {
      const handleSubmitMock = jest.fn();
      const excludeButton = getByTestId(container, 2)

      excludeButton.addEventListener('click', handleSubmitMock, false)

      await fireEvent.click(excludeButton)
      expect(handleSubmitMock).toHaveBeenCalled()
    })

    test('Should a list with 3 contact cards, when click a exclude button', async () => {
      const handleSubmitMock = jest.fn();
      const excludeButton = getByTestId(container, 2)

      excludeButton.addEventListener('click', handleSubmitMock, false)

      await fireEvent.click(excludeButton)
      expect(handleSubmitMock).toHaveBeenCalled()

      waitForElementToBeRemoved(() => excludeButton.parentNode)
      expect(container).toMatchSnapshot()
    })
  })

})

