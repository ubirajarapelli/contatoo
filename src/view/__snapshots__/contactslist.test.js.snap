// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests from a Contact List When a contact\`s list have contacts Should a list with 3 contact cards, when click a exclude button 1`] = `
<section
  id="contact-list"
>
  <article>
    <dl>
      <dt>
        Nome
      </dt>
      <dd
        contenteditable="true"
      >
        My name 1
      </dd>
      <dt>
        E-mail
      </dt>
      <dd
        contenteditable="true"
      >
        myemail1@test.com.br
      </dd>
      <dt>
        CPF
      </dt>
      <dd
        contenteditable="true"
      >
        04080757247
      </dd>
      <dt>
        Telefone
      </dt>
      <dd
        contenteditable="true"
      >
        11987654321
      </dd>
    </dl>
    <div>
      <button
        aria-label="Excluir contato"
        class="exclude"
        data-testid="0"
        id="04080757247"
      >
        <span>
          ✖
        </span>
      </button>
    </div>
  </article>
  <article>
    <dl>
      <dt>
        Nome
      </dt>
      <dd
        contenteditable="true"
      >
        My name 2
      </dd>
      <dt>
        E-mail
      </dt>
      <dd
        contenteditable="true"
      >
        myemail2@test.com.br
      </dd>
      <dt>
        CPF
      </dt>
      <dd
        contenteditable="true"
      >
        77797584192
      </dd>
      <dt>
        Telefone
      </dt>
      <dd
        contenteditable="true"
      >
        11987654321
      </dd>
    </dl>
    <div>
      <button
        aria-label="Excluir contato"
        class="exclude"
        data-testid="1"
        id="77797584192"
      >
        <span>
          ✖
        </span>
      </button>
    </div>
  </article>
  <article>
    <dl>
      <dt>
        Nome
      </dt>
      <dd
        contenteditable="true"
      >
        My name 3
      </dd>
      <dt>
        E-mail
      </dt>
      <dd
        contenteditable="true"
      >
        myemail3@test.com.br
      </dd>
      <dt>
        CPF
      </dt>
      <dd
        contenteditable="true"
      >
        45486737688
      </dd>
      <dt>
        Telefone
      </dt>
      <dd
        contenteditable="true"
      >
        11987654321
      </dd>
    </dl>
    <div>
      <button
        aria-label="Excluir contato"
        class="exclude"
        data-testid="2"
        id="45486737688"
      >
        <span>
          ✖
        </span>
      </button>
    </div>
  </article>
  <article>
    <dl>
      <dt>
        Nome
      </dt>
      <dd
        contenteditable="true"
      >
        My name 4
      </dd>
      <dt>
        E-mail
      </dt>
      <dd
        contenteditable="true"
      >
        myemail4@test.com.br
      </dd>
      <dt>
        CPF
      </dt>
      <dd
        contenteditable="true"
      >
        74668869066
      </dd>
      <dt>
        Telefone
      </dt>
      <dd
        contenteditable="true"
      >
        11987654321
      </dd>
    </dl>
    <div>
      <button
        aria-label="Excluir contato"
        class="exclude"
        data-testid="3"
        id="74668869066"
      >
        <span>
          ✖
        </span>
      </button>
    </div>
  </article>
</section>
`;

exports[`Unit tests from a Contact List When a load the contact\`s list page Should render a empty list 1`] = `
<section
  id="contact-list"
>
  <div>
    Você ainda não possui contatos ;)
  </div>
</section>
`;

exports[`Unit tests from a Contact List When a load the contact\`s list page Should render with a inital data 1`] = `
<section
  id="contact-list"
>
  <article>
    <dl>
      <dt>
        Nome
      </dt>
      <dd
        contenteditable="true"
      >
        My name 1
      </dd>
      <dt>
        E-mail
      </dt>
      <dd
        contenteditable="true"
      >
        myemail1@test.com.br
      </dd>
      <dt>
        CPF
      </dt>
      <dd
        contenteditable="true"
      >
        04080757247
      </dd>
      <dt>
        Telefone
      </dt>
      <dd
        contenteditable="true"
      >
        11987654321
      </dd>
    </dl>
    <div>
      <button
        aria-label="Excluir contato"
        class="exclude"
        data-testid="0"
        id="04080757247"
      >
        <span>
          ✖
        </span>
      </button>
    </div>
  </article>
  <article>
    <dl>
      <dt>
        Nome
      </dt>
      <dd
        contenteditable="true"
      >
        My name 2
      </dd>
      <dt>
        E-mail
      </dt>
      <dd
        contenteditable="true"
      >
        myemail2@test.com.br
      </dd>
      <dt>
        CPF
      </dt>
      <dd
        contenteditable="true"
      >
        77797584192
      </dd>
      <dt>
        Telefone
      </dt>
      <dd
        contenteditable="true"
      >
        11987654321
      </dd>
    </dl>
    <div>
      <button
        aria-label="Excluir contato"
        class="exclude"
        data-testid="1"
        id="77797584192"
      >
        <span>
          ✖
        </span>
      </button>
    </div>
  </article>
  <article>
    <dl>
      <dt>
        Nome
      </dt>
      <dd
        contenteditable="true"
      >
        My name 3
      </dd>
      <dt>
        E-mail
      </dt>
      <dd
        contenteditable="true"
      >
        myemail3@test.com.br
      </dd>
      <dt>
        CPF
      </dt>
      <dd
        contenteditable="true"
      >
        45486737688
      </dd>
      <dt>
        Telefone
      </dt>
      <dd
        contenteditable="true"
      >
        11987654321
      </dd>
    </dl>
    <div>
      <button
        aria-label="Excluir contato"
        class="exclude"
        data-testid="2"
        id="45486737688"
      >
        <span>
          ✖
        </span>
      </button>
    </div>
  </article>
  <article>
    <dl>
      <dt>
        Nome
      </dt>
      <dd
        contenteditable="true"
      >
        My name 4
      </dd>
      <dt>
        E-mail
      </dt>
      <dd
        contenteditable="true"
      >
        myemail4@test.com.br
      </dd>
      <dt>
        CPF
      </dt>
      <dd
        contenteditable="true"
      >
        74668869066
      </dd>
      <dt>
        Telefone
      </dt>
      <dd
        contenteditable="true"
      >
        11987654321
      </dd>
    </dl>
    <div>
      <button
        aria-label="Excluir contato"
        class="exclude"
        data-testid="3"
        id="74668869066"
      >
        <span>
          ✖
        </span>
      </button>
    </div>
  </article>
</section>
`;
